examples/examples/rpc_middleware_rate_limiting.rs

```diff
@@ -78,54 +78,73 @@ impl<S> RateLimit<S> {
 			state: Arc::new(Mutex::new(State::Allow { until: Instant::now() + period, rem: num + 1 })),
 		}
 	}
-}
 
-impl<'a, S> RpcServiceT<'a> for RateLimit<S>
-where
-	S: Send + RpcServiceT<'a>,
-{
-	// Instead of `Boxing` the future in this example
-	// we are using a jsonrpsee's ResponseFuture future
-	// type to avoid those extra allocations.
-	type Future = ResponseFuture<S::Future>;
-
-	fn call(&self, req: Request<'a>) -> Self::Future {
+	fn rate_limit_deny(&self) -> bool {
 		let now = Instant::now();
-
-		let is_denied = {
-			let mut lock = self.state.lock().unwrap();
-			let next_state = match *lock {
-				State::Deny { until } => {
-					if now > until {
-						State::Allow { until: now + self.rate.period, rem: self.rate.num - 1 }
-					} else {
-						State::Deny { until }
-					}
+		let mut lock = self.state.lock().unwrap();
+		let next_state = match *lock {
+			State::Deny { until } => {
+				if now > until {
+					State::Allow { until: now + self.rate.period, rem: self.rate.num - 1 }
+				} else {
+					State::Deny { until }
 				}
-				State::Allow { until, rem } => {
-					if now > until {
-						State::Allow { until: now + self.rate.period, rem: self.rate.num - 1 }
-					} else {
-						let n = rem - 1;
-						if n > 0 {
-							State::Allow { until: now + self.rate.period, rem: n }
-						} else {
-							State::Deny { until }
-						}
-					}
+			}
+			State::Allow { until, rem } => {
+				if now > until {
+					State::Allow { until: now + self.rate.period, rem: self.rate.num - 1 }
+				} else {
+					let n = rem - 1;
+					if n > 0 { State::Allow { until: now + self.rate.period, rem: n } } else { State::Deny { until } }
 				}
-			};
-
-			*lock = next_state;
-			matches!(next_state, State::Deny { .. })
+			}
 		};
 
-		if is_denied {
+		*lock = next_state;
+		matches!(next_state, State::Deny { .. })
+	}
+}
+
+impl<S> RpcServiceT for RateLimit<S>
+where
+	S: Send + RpcServiceT<Response = MethodResponse> + 'static,
+	S::Error: Send,
+{
+	type Error = S::Error;
+	type Response = S::Response;
+
+	fn call<'a>(&self, req: Request<'a>) -> impl Future<Output = Result<Self::Response, Self::Error>> + Send + 'a {
```

Comment:
I haven't tried Rust 2024 yet but it has this new `use` syntax, so I wonder whether it's more correct to have `+ use<'a>` in the response rather than `+ 'a`? I always forget the subtleties around why `+ 'a` isn't quite the same thing!

Reply:
we need https://github.com/rust-lang/rust/pull/138128 for that to work with impl Trait<Self::Foo>

Reply:
Ooh I didn't realise it wasn't stable for traits yet, but yay for that PR having merged!